### ISSUE TITLE
Cannot define declarative base class as attribute of object

### ISSUE THREAD

**Issue by MaicoTimmerman** [sqlalchemy/sqlalchemy2-stubs#54]
**Describe the bug**
I'd like to switch from the [dropbox/sqlalchemy](https://github.com/dropbox/sqlalchemy-stubs) stubs to the official stubs, however when the declarative base is defined as attribute of an instance, it is not recognized as valid base class.

The Dropbox stubs give similar errors regarding the subclassing, however the types are still correctly inferred.

**Expected behavior**
Type definitions to be picked up as the builtin types: `test.py:20: note: Revealed type is 'Union[builtins.int, None]'`

**To Reproduce**
```py
from sqlalchemy.schema import Column
from sqlalchemy.types import Integer, String
from sqlalchemy.orm.decl_api import DeclarativeMeta, declarative_base

db = DB()

class DB():
    def __init__(self):
        self.Base = declarative_base()


class A(db.Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)

a = A()
reveal_type(a.id)
```

**Error**

```
test2.py:12: error: Name 'db.Base' is not defined
test2.py:17: note: Revealed type is 'sqlalchemy.sql.schema.Column[sqlalchemy.sql.sqltypes.Integer*]'
Found 1 error in 1 file (checked 1 source file)
```

**Versions.**
 - OS: Arch linux
 - Python: 3.9
 - SQLAlchemy: 1.4.5 + mypy plugin
 - Database: n/a
 - DBAPI: n/a

**Additional context**
The problem seems similar to one described [on StackOverflow](https://stackoverflow.com/questions/63542818/mypy-and-inheriting-from-a-class-that-is-an-attribute-on-an-instance). However, I expected that replacing the direct subclassing with a typed variable would resort the problem.

<details>
<summary>Example code</summary>

```py
from sqlalchemy.schema import Column
from sqlalchemy.types import Integer, String
from sqlalchemy.orm.decl_api import DeclarativeMeta, declarative_base

db = DB()

class DB():
    def __init__(self):
        self.Base = declarative_base()

Base: DeclarativeMeta = db.Base

class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)

a = A()
reveal_type(a.id)
```

</details>

**Have a nice day!**

**Comment by zzzeek** [comment]
your example refers to "DB" before the class is declared, which would be part of the problem.    However, I can't get this pattern to work at all for Mypy, even without any dynamic classes or anything:

```python

class Foo:
    pass


class DB:
    def __init__(self):
        self.Base = Foo


db = DB()


class A(db.Base):
    pass

```

```
$ mypy test3.py 
test3.py:13: error: Name 'db.Base' is not defined
Found 1 error in 1 file (checked 1 source file)

```

not sure what I'm missing? or is this a known mypy bug?

**Comment by zzzeek** [comment]
oh, as far as why that prevents everything else from working, the SQLAlchemy stubs rely upon the [mypy plugin](https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html?highlight=mypy) for declarative mappings to be supported.  The plugin can't work here because mypy is not giving the plugin any information about declarative_base() here because as above Mypy seems to be broken for this case.

**Comment by layday** [comment]
It's generally the case that the return type of a callable can't be used in typing contexts.  Mypy should probably print something more informative than "Name 'db.Base' is not defined" here.

**Comment by MaicoTimmerman** [comment]
I think this might fall under the unsupported python features: _Run-time evaluated base classes like class A(foo()): pass, Similar to metaclasses_ [1]. Currently, there is support for detecting the `declarative_base()` function directly and replace it with a full blown type including mro, perhaps that could be expanded to also detect it in functions and classes? Like:

```py
from sqlalchemy.schema import Column
from sqlalchemy.types import Integer
from sqlalchemy.orm.decl_api import declarative_base

def my_declarative_base():
    return declarative_base()

Base = my_declarative_base()

class A(Base):
    __tablename__ = "a"
    id = Column(Integer, primary_key=True)

a = A()
reveal_type(a.id)
```



[1] https://github.com/python/mypy/wiki/Unsupported-Python-Features

**Comment by zzzeek** [comment]
if you have a standalone Base, the mypy plugin supports that just fine through the existing plugin hooks.

Also I would recommend, if you truly have "decalrative bases per instance of X" going on, use the [decorator pattern](https://docs.sqlalchemy.org/en/14/orm/mapping_styles.html#declarative-mapping-using-a-decorator-no-declarative-base), where the "registry" is the attribute of your object that contains "base".   Mypy seems to be more forgiving about this kind of thing (but maybe not in this case still, not sure)

**Comment by synic** [comment]
This definitely seems more like a problem with mypy, and not anything to do with the sqlalchemy-stubs:

```python
from typing import Type


class FakeBaseModel:
    pass


class FakeDatabase:
    Base: Type[FakeBaseModel]

    def __init__(self) -> None:
        self.Base = FakeBaseModel


db = FakeDatabase()


class FakeModel(db.Base):
    pass
```

```bash
[mypytest {mypytest}]$ mypy .
mypytest/models.py:18: error: Name "db.Base" is not defined
Found 1 error in 1 file (checked 2 source files)
```

From https://stackoverflow.com/questions/71069074/appeasing-mypy-trying-to-use-an-instance-variable?noredirect=1#comment125650705_71069074

**Comment by zzzeek** [comment]
yes I had a convo with the typing folks and functions that return classes are a no-go in general, none of the tools support it
